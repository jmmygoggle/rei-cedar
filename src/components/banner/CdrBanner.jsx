import clsx from 'clsx';
import propValidator from '../../utils/propValidator';
import BuildClass from '../../mixins/buildClass';
import style from './styles/CdrBanner.scss';

export default {
  name: 'CdrBanner',
  mixins: [BuildClass],
  props: {
    type: {
      type: String,
      validator: (value) => propValidator(
        value,
        ['info', 'warning', 'success', 'error'],
      ),
      default: 'info',
    },
  },
  data() {
    return {
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-banner';
    },
    borderClass() {
      return this.modifyClassName(this.baseClass, this.type);
    },
    iconClass() {
      return this.modifyClassName(`${this.baseClass}__icon-left`, this.type);
    },
    mainTheme() {
      return this.modifyClassName(`${this.baseClass}__main`, this.type);
    },
    prominenceClass() {
      return this.$slots['message-body']
        ? this.modifyClassName(`${this.baseClass}__wrapper`, 'prominence')
        : undefined;
    },
    noMsgBodyClass() {
      return !this.$slots['message-body']
        ? this.modifyClassName(`${this.baseClass}__main`, 'no-msg')
        : undefined;
    },
  },
  render() {
    return (
      <div class={clsx(this.style[this.baseClass], this.borderClass)}>
        <div class={clsx(this.style['cdr-banner__wrapper'], this.prominenceClass)}>
          <div class={clsx(this.style['cdr-banner__main'], this.mainTheme, this.noMsgBodyClass)}>
            {this.$slots['icon-left'] && (
              <div class={clsx(this.style['cdr-banner__icon-left'], this.iconClass)}>
                {this.$slots['icon-left']}
              </div>
            )}
            <span class={this.style['cdr-banner__message']}>
              { this.$slots.default }
            </span>
            {this.$slots['icon-right'] && (
              <div class={this.style['cdr-banner__icon-right']}>
                {this.$slots['icon-right']}
              </div>
            )}
          </div>
          {this.$slots['message-body'] && (
            <div class={this.style['cdr-banner__message-body']}>
              {this.$slots['message-body']}
            </div>
          )}
        </div>
        {this.$slots['info-action'] && (
          <div class={this.style['cdr-banner__info-action']}>
            {this.$slots['info-action']}
          </div>
        )}
      </div>
    );
  },
};
