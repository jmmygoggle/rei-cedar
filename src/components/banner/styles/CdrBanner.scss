@import '../../../css/settings/index.scss';
@import './CdrBanner.vars.scss';

.cdr-banner {
  @include cdr-banner-base-mixin;
  display: grid;
  grid-template-columns: 1fr 3.2rem;
  grid-template-areas: 
  "main info-action"
  "message-body info-action"
  ;
  &__wrapper {
    border-radius: $cdr-radius-softer;
    &--prominence {
      box-shadow: $cdr-prominence-raised;
    }
  }
  &__main {
    grid-area: main;
    display: grid;
    border-top-right-radius: $cdr-radius-softer;
    grid-template-columns: auto 1fr auto;
    grid-template-areas: "icon-left message icon-right";
    &--no-msg {
      border-bottom-right-radius: $cdr-radius-softer;
    }
  }
  &__icon-left {
    grid-area: icon-left;
    display: inherit;
    align-items: center;
    justify-items: center;
    width: 4rem;
    height: 4rem;
    & svg {
      width: 3.2rem;
      stroke: white;
      stroke-width: 0.4rem;
      paint-order: stroke fill;
      margin: $cdr-space-quarter-x;
  
    }
  }
  &__message {
    grid-area: message;
    padding: $cdr-space-half-x;
  }
  &__message-body {
    grid-area: message-body;
    border-bottom-right-radius: $cdr-radius-softer;
    background-color: $cdr-color-background-primary;
    padding: $cdr-space-half-x;
  }
  &__icon-right {
    grid-area: icon-right;
    & svg {
      fill: $cdr-color-text-emphasis!important;
    }
  }
  &__info-action {
    grid-area: info-action;
    & svg {
      fill: $cdr-color-text-link-rest!important;
    }
  }
  &--default {
    & .cdr-banner__main {
      @include cdr-banner-default-mixin;
      & .cdr-banner__icon-left {
          background-color: $cdr-color-background-message-default-02;
        }
      }
    border-left-color: $cdr-color-border-message-default-01;
  }

  &--info {
    & .cdr-banner__main {
      @include cdr-banner-info-mixin;
      & .cdr-banner__icon-left {
        background-color: $cdr-color-background-message-info-02;
      }
    }
    border-left-color: $cdr-color-border-message-info-01;
  }

  &--warning {
    & .cdr-banner__main {
      @include cdr-banner-warning-mixin;
      & .cdr-banner__icon-left {
        background-color: $cdr-color-background-message-warning-02;
      }
    }
    border-left-color: $cdr-color-border-message-warning-01;
  }

  &--success {
    & .cdr-banner__main {
      @include cdr-banner-success-mixin;
      & .cdr-banner__icon-left {
        background-color: $cdr-color-background-message-success-02;
      }
    }
    border-left-color: $cdr-color-border-message-success-01;
  }

  &--error {
    & .cdr-banner__main {
      @include cdr-banner-error-mixin;
      & .cdr-banner__icon-left {
        background-color: $cdr-color-background-message-error-02;
      }
    }
    border-left-color: $cdr-color-border-message-error-01;
  }
}

