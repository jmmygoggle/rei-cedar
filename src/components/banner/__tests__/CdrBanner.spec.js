import { mount } from '@vue/test-utils';
import CdrBanner from 'componentdir/banner/CdrBanner';
import IconCheckFill from 'componentdir/icon/comps/check-fill';
describe('CdrBanner', () => {
  it('renders correctly', () => {
    const wrapper = mount(CdrBanner, {
      propsData: {
        type: 'info',
      },
      slots: {
       default: 'hey im a banner',
       'icon-left': IconCheckFill
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('renders a right-icon', () => {
    const wrapper = mount(CdrBanner, {
      propsData: {
        type: 'warning',
      },
      slots: {
       default: 'hey im a banner',
       'icon-left': IconCheckFill,
       'icon-right': IconCheckFill,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('renders the info-action slot', () => {
    const wrapper = mount(CdrBanner, {
      propsData: {
        type: 'warning',
      },
      slots: {
       default: 'hey im a banner',
       'icon-left': IconCheckFill,
       'icon-right': IconCheckFill,
       'info-action': IconCheckFill,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('renders the message-body slot', () => {
    const wrapper = mount(CdrBanner, {
      propsData: {
        type: 'success',
      },
      slots: {
       default: 'hey im a banner',
       'icon-left': IconCheckFill,
       'message-body': 'and I am some extra information'
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
