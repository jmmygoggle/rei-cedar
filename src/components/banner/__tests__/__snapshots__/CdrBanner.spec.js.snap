// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrBanner renders a right-icon 1`] = `
<div
  class="cdr-banner cdr-banner--warning"
>
  <div
    class="cdr-banner__wrapper"
  >
    <div
      class="cdr-banner__main cdr-banner__main--warning cdr-banner__main--no-msg"
    >
      <div
        class="cdr-banner__icon-left cdr-banner__icon-left--warning"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm4.77-13.357a1 1 0 10-1.533-1.286l-5.95 7.083-1.52-1.817a1 1 0 00-1.533 1.285l2.227 2.66a.999.999 0 001.565.1l6.743-8.025z"
            role="presentation"
          />
        </svg>
      </div>
      <span
        class="cdr-banner__message"
      >
        hey im a banner
      </span>
      <div
        class="cdr-banner__icon-right"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm4.77-13.357a1 1 0 10-1.533-1.286l-5.95 7.083-1.52-1.817a1 1 0 00-1.533 1.285l2.227 2.66a.999.999 0 001.565.1l6.743-8.025z"
            role="presentation"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`CdrBanner renders correctly 1`] = `
<div
  class="cdr-banner cdr-banner--info"
>
  <div
    class="cdr-banner__wrapper"
  >
    <div
      class="cdr-banner__main cdr-banner__main--info cdr-banner__main--no-msg"
    >
      <div
        class="cdr-banner__icon-left cdr-banner__icon-left--info"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm4.77-13.357a1 1 0 10-1.533-1.286l-5.95 7.083-1.52-1.817a1 1 0 00-1.533 1.285l2.227 2.66a.999.999 0 001.565.1l6.743-8.025z"
            role="presentation"
          />
        </svg>
      </div>
      <span
        class="cdr-banner__message"
      >
        hey im a banner
      </span>
    </div>
  </div>
</div>
`;

exports[`CdrBanner renders the info-action slot 1`] = `
<div
  class="cdr-banner cdr-banner--warning"
>
  <div
    class="cdr-banner__wrapper"
  >
    <div
      class="cdr-banner__main cdr-banner__main--warning cdr-banner__main--no-msg"
    >
      <div
        class="cdr-banner__icon-left cdr-banner__icon-left--warning"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm4.77-13.357a1 1 0 10-1.533-1.286l-5.95 7.083-1.52-1.817a1 1 0 00-1.533 1.285l2.227 2.66a.999.999 0 001.565.1l6.743-8.025z"
            role="presentation"
          />
        </svg>
      </div>
      <span
        class="cdr-banner__message"
      >
        hey im a banner
      </span>
      <div
        class="cdr-banner__icon-right"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm4.77-13.357a1 1 0 10-1.533-1.286l-5.95 7.083-1.52-1.817a1 1 0 00-1.533 1.285l2.227 2.66a.999.999 0 001.565.1l6.743-8.025z"
            role="presentation"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="cdr-banner__info-action"
  >
    <svg
      aria-hidden="true"
      class="cdr-icon"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm4.77-13.357a1 1 0 10-1.533-1.286l-5.95 7.083-1.52-1.817a1 1 0 00-1.533 1.285l2.227 2.66a.999.999 0 001.565.1l6.743-8.025z"
        role="presentation"
      />
    </svg>
  </div>
</div>
`;

exports[`CdrBanner renders the message-body slot 1`] = `
<div
  class="cdr-banner cdr-banner--success"
>
  <div
    class="cdr-banner__wrapper cdr-banner__wrapper--prominence"
  >
    <div
      class="cdr-banner__main cdr-banner__main--success"
    >
      <div
        class="cdr-banner__icon-left cdr-banner__icon-left--success"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm4.77-13.357a1 1 0 10-1.533-1.286l-5.95 7.083-1.52-1.817a1 1 0 00-1.533 1.285l2.227 2.66a.999.999 0 001.565.1l6.743-8.025z"
            role="presentation"
          />
        </svg>
      </div>
      <span
        class="cdr-banner__message"
      >
        hey im a banner
      </span>
    </div>
    <div
      class="cdr-banner__message-body"
    >
      and I am some extra information
    </div>
  </div>
</div>
`;
